### PR TITLE
ENH: Developer CLI for SciPy

### PR DESCRIPTION
**Pull request by sayantikabanik**
#### Reference issue

- Part of https://github.com/scipy/scipy/issues/15489
- POC details https://github.com/rgommers/scipy/issues/133

#### What does this implement/fix?

- Implements a developer CLI using `doit`, `click` along with a GUI implemented using `rich-click`. 

#### Handy commands 
- Enabling the GUI: `python do.py`
- Listing the args/options for a task: `python do.py <task_name> --help`

#### Additional information

*Current list of implemented tasks:*

* `Build & testing tasks`

     - build (build & install package on path)
     - test (Run tests along with options to run tests for a given submodule)
* `Static checker tasks`

     - pep8 ( Perform pep8 check with flake8)
     - mypy ( Run mypy on the codebase)
* `Environments`

     - shell (Start Unix shell with PYTHONPATH set)
     - python (Start a Python shell with PYTHONPATH set)
     - ipython (Start IPython shell with PYTHONPATH set )
* `Documentation tasks`

     - doc (Build documentation)
     - refguide-check  (Run refguide check)
 
* `Release tasks`
     
    -  notes   (Release notes and log generation)                                                                                                                                                                                
    -  authors  (Task to generate list the authors who contributed within a given revision interval)
  
* `Benchmarking tasks`
    - bench & bench compare

#### GUI look 
Screenshot 2022-04-08 at 9 08 43 PM

cc @rgommers @schettino72

### CODE REVIEW

**Comment by tupui** [comment]
The GUI looks awesome! Great work 👏

File: environment.yml [hunk 1]
```diff
@@ -38,3 +38,8 @@ dependencies:
   - mpmath
   - gmpy2
   - threadpoolctl
+  # For CLI
+  - rich-click
+  - click
+  - pip:
```
Comment (tupui) [inline environment.yml:44]: We an already remove this I think since the release on conda is in progress.

File: do.py [hunk 2]
```diff
@@ -0,0 +1,1174 @@
...
+
+- [ ] doc: it would make sense to expose MAKE targets as sub-commands or options
+
+cleanup + doit:
+- [ ] move out non-scipy code
+- [ ] doit reporter when running under click
+
+BUG:
+- [ ] python dev.py -t scipy.optimize.tests.test_minimize_constrained.py::TestTrustRegionConstr::test_args
+      unknown marker "slow". seems conftest is not being loaded
```
Comment (tupui) [inline do.py:107]: I thought this was solved with latest pytest. They had a regression on 7.0 and 7.0.1 should fix it.

**Review by rgommers — COMMENTED** [review]
Thanks @sayantikabanik and @schettino72 for all the hard work to put together this developer interface - I think it will be a major upgrade to the experience of working on SciPy!

@scipy/maintainers this isn't 100% final but we've been using and tweaking it for a little while, and it's ready for merging now imho. Smaller things can be added/fixed in follow-ups. I'll add here the current plan for this CLI, and what I think are the main things to look at (aside from just trying the workflows you care about and playing around) for review.

The plan:
1. Merge this PR in its current state (modulo review comments), with experimental status - so no docs yet, and reserve the freedom to make some changes over the next weeks.
2. There's still a part of this in `do.py` (clearly marked) that @schettino72 wants to move into either `doit` or a separate package. That will simplify maintenance of this file, and make it reusable by other projects.
3. Once we're happy with this CLI, remove `dev.py` and rename `do.py` to `dev.py` (better name, and `do` is a bash builtin).
4. Update contributor docs for this (note that a lot of docs are in `--help` for each command, so we need fewer html docs).
5. After 1.9.0, we can remove `numpy.distutils` build infra, and hence also `runtests.py`. At that point we have a single CLI again, this one.

To review:
- The plan above
- The added dependencies. They're not too heavy, but there are a few new ones. I think the screenshot in the PR description already shows that that's worth it.
- The complexity of the logic, for `build`, `test` and `bench` tasks in particular. I think it's an improvement over the massive if-elif-else block in `dev.py`/`runtests.py`. But maybe there's room for further improvement.

**Review by tupui — COMMENTED** [review]
I agree with the plan, we should not spend time on nitpicks there. Provided the added commands are working, I am fine with getting it in to iterate faster. If we start using this in CI, we will see rapidly what is working and what not.

On `rich` itself, I am neutral. It does look nicer, but it's not essential and `rich` is still moving very fast. But now that `pip` started to use it, it might be a sign of confidence/stability/guarantee.

File: do.py [hunk 3]
```diff
@@ -0,0 +1,1174 @@
...
+    def task_meta(cls, **kwargs):
+        return {
+```
+
+
+
+# TODO:
+First milestone replace dev.py
+
+- [ ] doc: it would make sense to expose MAKE targets as sub-commands or options
```
Comment (rgommers) [inline do.py:99]: Note that I plan to seriously cull what's in that Makefile. We really only need `html`, `show`, `upload`, `dist`. That's for another time though, for now what's in the CLI is enough.

File: do.py (hunk 2 above)
Comment (schettino72) [reply do.py:107]: It still fails for me on pytest 7.1.1
Note the command line above reproduces an error on existing `dev.py`, not on the code in this PR.

**Comment by schettino72** [comment]
Thanks @rgommers and @sayantikabanik for support and collaboration.

I would just like to add to the plan (at least from my side and if you guys agree).
This PR has improvements for the CLI itself...

But now since commands tasks are based on `doit`, there is room for improvements on individual tasks.
Mostly speed up by taking advantage of up-to-date checks and incremental execution.

**Comment by rgommers** [comment]
> But now since commands tasks are based on `doit`, there is room for improvements on individual tasks.
> Mostly speed up by taking advantage of up-to-date checks and incremental execution.

This would be interesting in principle, anything where we can improve the lag on the development cycle will be quite helpful. The question is which commands though. Right now, every tool has its own cache, at least these ones:
- Meson (build) 
- Sphinx (doc build)
- ASV (benchmarks, and build envs for `--compare`)
- Mypy (type annotations, a bit flaky IIRC)

The Meson cache is reliable and fast, and `doit` doesn't have enough info to skip invoking Meson. For ASV there may be an opportunity to skip the "environment up to date" checks, not sure. Sphinx is glacially slow, but it's all a single step, so not sure what `doit` can do about that. Mypy the same I think.

Perhaps the bigger gain here would be for composite commands, like "run through the whole release workflow for tag `v1.8.0`". Thinking about that some more, it's the composability of tasks more than skipping tasks that will be helpful. The `dev.py` style if-else logic make it quite hard to extend. Now we can more easily define new tasks that depend on other tasks. Those multi-task jobs where `doit` is in control are then in turn much easier to ensure "task is up-to-date and hence can be skipped" is feasible, because there's no environment activation, change in build configuration, etc. that can mess things up in between two tasks.

File: environment.yml (hunk 1 above)
Comment (sayantikabanik) [reply environment.yml]: Updated, though for `arm64` it failed with `PackagesNotFoundError:`. It's not yet available under `conda-forge/osx-arm64` or `conda-forge/noarch
`

Comment (rgommers) [reply environment.yml]: Opened an issue for that at https://github.com/conda-forge/doit-feedstock/issues/16

Comment (rgommers) [reply environment.yml]: Looks like we won't get that arm64 package, so let's revert back to `pip` until there's a new `doit` release which is `noarch`.

**Comment by rgommers** [comment]
Hi all, I plan to merge this by Friday unless there's blocking concerns. That will help get people to use it and find possible parts that aren't working smoothly just yet. We'll keep `dev.py` around until we're sure this is all good to go.

File: do.py (hunk 2 above)
Comment (sayantikabanik) [reply do.py:107]: Failed for me as well using `pytest 7.7.1`

```shell
pytest.PytestUnknownMarkWarning: Unknown pytest.mark.slow 
```

Comment (schettino72) [reply do.py:107]: @sayantikabanik feel free to look into this. I am busy with other stuff...

**Comment by grlee77** [comment]
This looks great! I tried  a few of the commands and things were working as expected. I tried the `test`, `doc` and `pep8` commands and they ran as expected. The only issue I noticed is that `python do.py pep8` does result in a lot of flake8 warnings about the new `do.py` itself.

**Comment by schettino72** [comment]
@grlee77 I have just started to work on fixing flake8 issues. I should be able to push an update today.

**Comment by schettino72** [comment]
I fixed the pep8 issues in `do.py`.

Minor issue: I guess we should take this opportunity to rename the command `pep8`.
`pep8` package was renamed to `pycodestyle`...  
And tool used is actually `flake8` which uses executes `pycodestyle` and `pyflakes` internally.

**Comment by rgommers** [comment]
> Minor issue: I guess we should take this opportunity to rename the command `pep8`.

Yes you're right, I'd suggest we call it `lint` because what it does is linting in the same way as we do in CI (where it's also called lint). It actually does two separate things:
1. Run `flake8` over the whole code base
2. Check for PEP 8 compliance on the diff between the branch and `main` by running the `tools/lint_diff.py` script.

I'll push a change.

**Comment by rgommers** [comment]
I added a new emoji for "doesn't depend on the `build` task" to make sure things line up.

**Review by rgommers — APPROVED** [review]
> Hi all, I plan to merge this by Friday unless there's blocking concerns. That will help get people to use it and find possible parts that aren't working smoothly just yet. We'll keep `dev.py` around until we're sure this is all good to go.

It is now Friday, and this has been tested by a fair number of people. So let's merge this as is. A huge thank you to @schettino72 and @sayantikabanik for doing the heavy lifting. And of course to everyone else who contributed ideas/testing/review.

There are a few follow-ups listed in the file itself. Others I'm aware of:
- clean up docs Makefile
- add support for `act` in `do.py` (@sayantikabanik has changes ready for that)
- once the `doit` conda package is installable as `noarch`, update the environment yml files
- look at new opportunities provided by the new structure (caching w `doit`, more complex workflows, ...?)

And one that just occurred to me and may make sense:
- make the `--help` output paging, because it's too long to fit on the average terminal.

File: do.py (hunk 2 above)
Comment (schettino72) [reply do.py:104]: :facepalm: The problem is the example is using python module notation (separated by dot), but also includes `.py`.
A pytest bug is responsible for the misleading error message...

WRONG:
```
python dev.py -t scipy.optimize.tests.test_minimize_constrained.py::TestTrustRegionConstr::test_args
```

CORRECT:
```
python dev.py -t scipy.optimize.tests.test_minimize_constrained::TestTrustRegionConstr::test_args
python dev.py -t scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_args
```

**Comment by sayantikabanik** [comment]
> * add support for `act` in `do.py` (@sayantikabanik has changes ready for that)

Currently facing an issue which is also reported under https://github.com/scipy/scipy/issues/16009
```python
[macOS tests (meson)/Meson build] Skipping job 'Meson build' due to 'github.repository == 'scipy/scipy' || github.repository == '''
```

**Comment by AnirudhDagar** [comment]
When there is any kind of error, it is not logged on the terminal. Instead, the script just exits without telling anything.

### How to reproduce

Just add any random erroneous line to do.py file task. https://github.com/scipy/scipy/blob/9137ef1c201aa75f891fe2ad5ac056ad9d2ff5d4/do.py#L545-L551

 Here I'm adding the following before line 550.

```python
     @classmethod 
     def build_project(cls, dirs, args, env): 
         """ 
         Build a dev version of the project. 
         """ 
         import thisdoesnotexist  #obviously this should error out
         cmd = ["ninja", "-C", str(dirs.build)] 
         if args.parallel > 1:
```

But when I run `python do.py build`, the script just runs and exits at this line without any error or traceback output.
Is this issue already known since do.py is under development atm? Asking here quickly, otherwise, I'll open up an issue.

cc @sayantikabanik @schettino72 @rgommers

**Comment by AnirudhDagar** [comment]
nevermind, just found this comment https://github.com/rgommers/scipy/issues/133#issuecomment-1106647572

Apologies for the noise.

**Pull request merged**